### PR TITLE
Mike & Lucas review the world.

### PR DESCRIPTION
**Pull request by ltw**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,3 +14,5 @@
 # Ignore all logfiles and tempfiles.
 /log/*.log
 /tmp
+.DS_Store
```
Comment (ltw) [inline .gitignore]: to stop adding this to every project gitignore you work on, consider adding this to a global gitignore.

File: Gemfile [hunk 2]
```diff
@@ -27,14 +28,30 @@ gem 'sdoc', '~> 0.4.0',          group: :doc
...
 # gem 'capistrano-rails', group: :development
 
 # Use debugger
 # gem 'debugger', group: [:development, :test]
 
+# Use factory girl
+gem 'factory_girl_rails'
+
+#Use rails_12factor
+gem 'rails_12factor'
```
Comment (ltw) [inline Gemfile]: this should be in a production group

File: app/controllers/questions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  def new
+    @question = Question.new
+  end
+
+  def create
+    @question = Question.new(question_params)
+    if @question.save
+      @question.set_posted_at
```
Comment (ltw) [inline app/controllers/questions_controller.rb]: This should be implemented as a `before_create`

File: app/models/question.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  has_many :tags, through: :question_tags
+  has_many :votes, as: :votable
+  has_many :responses, as: :respondable
+
+  validates :title, :content, :user_id, presence: true
+  validates :title, uniqueness: true
+  validates :title, length: { maximum: 90 }
+  validates :content, length: { maximum: 10000 }
+
+  def set_posted_at
```
Comment (ltw) [inline app/models/question.rb]: Do you need `posted_at` when it's only ever going to be set to `created_at`?

File: app/models/question.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    self.save
+  end
+
+  def posted_at
+    super.strftime("on %m/%d/%Y at %I:%M%p")
+  end
+
+  def first(number)
+    if self.content.length > number
+      (self.content[0..number] + " <a href='/questions/#{self.id}'>...continue reading</a>").html_safe
```
Comment (ltw) [inline app/models/question.rb]: If you're constructing HTML, you should be doing it in the helpers, not the model.

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -1,14 +1,18 @@
...
 <html>
 <head>
-  <title>Metaoverflow</title>
+  <title><%= yield(:title) %></title>
   <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track' => true %>
   <%= javascript_include_tag 'application', 'data-turbolinks-track' => true %>
   <%= csrf_meta_tags %>
 </head>
 <body>
+  <header>
```
Comment (ltw) [inline app/views/layouts/application.html.erb]: Maybe have a header partial?

File: app/views/users/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  </div>
+  <%= @user.bio %>
+</div>
+
+<div class = "user-questions">
+  <%= @user.questions.count%>
+  <ul>
+  <% @user.questions.each do |question|%>
+    <li><%= question.title %>
+    <%= question.votes.count %></li>
```
Comment (ltw) [inline app/views/users/show.html.erb]: Suggestion: investigate a `counter_cache` column. Search for `counter_cache` here: http://guides.rubyonrails.org/association_basics.html#belongs-to-association-reference

File: config/newrelic.yml [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+# the environment.
+
+# NOTE if your application has other named environments, you should
+# provide newrelic configuration settings for these environments here.
+
+development:
+  <<: *default_settings
+  # Turn on communication to New Relic service in development mode
+  monitor_mode: true
+  app_name: My Application (Development)
```
Comment (ltw) [inline config/newrelic.yml]: You need to configure your New Relic application here.

File: config/routes.rb [hunk 9]
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
+  root 'questions#index'
```
Comment (ltw) [inline config/routes.rb]: `root` goes at the bottom because the routes file is hierarchical from top to bottom.

File: config/routes.rb [hunk 10]
```diff
@@ -1,4 +1,8 @@
...
+
+  resources :questions, :answers, :responses
```
Comment (ltw) [inline config/routes.rb]: I wouldn't inline these, as you will almost always want to use `only` or `except` to restrict access to the routes that you accept for each resource.

File: db/migrate/20150114194348_create_questions.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class CreateQuestions < ActiveRecord::Migration
+  def change
+    create_table :questions do |t|
+      t.string :title
+      t.text :content
+      t.datetime :posted_at
+      t.belongs_to :user
```
Comment (ltw) [inline db/migrate/20150114194348_create_questions.rb]: foreign key index required.

File: db/migrate/20150114194931_create_answers.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+class CreateAnswers < ActiveRecord::Migration
+  def change
+    create_table :answers do |t|
+      t.belongs_to :question
```
Comment (ltw) [inline db/migrate/20150114194931_create_answers.rb]: fk indexes

File: db/migrate/20150114195517_create_votes.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+class CreateVotes < ActiveRecord::Migration
+  def change
+    create_table :votes do |t|
+      t.belongs_to :user
```
Comment (ltw) [inline db/migrate/20150114195517_create_votes.rb]: fk index

File: db/migrate/20150114201048_create_responses.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
+class CreateResponses < ActiveRecord::Migration
+  def change
+    create_table :responses do |t|
+      t.belongs_to :user
```
Comment (ltw) [inline db/migrate/20150114201048_create_responses.rb]: fk index

File: db/migrate/20150114215018_create_question_tags.rb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+class CreateQuestionTags < ActiveRecord::Migration
+  def change
+    create_table :question_tags do |t|
+      t.belongs_to :question
```
Comment (ltw) [inline db/migrate/20150114215018_create_question_tags.rb]: fk indexes

File: db/seeds.rb [hunk 16]
```diff
@@ -1,7 +1,85 @@
...
+respondables += responses
+
+# create votes for everything that should be voted on
+respondables.each do |item|
+  rand(1..5).times do
+    item.votes.create!(user: users.sample)
+  end
+end
+
+# 10.times do
```
Comment (ltw) [inline db/seeds.rb]: don't commit commented out stuff.

File: spec/controllers/questions_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+
+RSpec.describe QuestionsController, :type => :controller do
+  describe "Get #index" do
+    before {get :index}
```
Comment (ltw) [inline spec/controllers/questions_controller_spec.rb]: Remember the rspec lecture - Given, When, Then. 
`before` is the Given, `it` is the When and Then. 
`get :index` is the When.

File: spec/factories/answers.rb [hunk 18]
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
+  factory :answer do
```
Comment (ltw) [inline spec/factories/answers.rb]: None of your factories have content. Not seeing the point of these.

File: spec/models/user_spec.rb [hunk 19]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe User, :type => :model do
+  it { should validate_presence_of(:username) }
```
Comment (ltw) [inline spec/models/user_spec.rb]: Good use of shoulda matchers.

**Comment by ltw** [comment]
Great beginnings. Couple of points:

1. Specs, specs, specs, specs, specs. More testing, more yay. The ones you have are awesome - keep it up.
2. More indexing on foreign keys. Also remember to add composite indexes for things that will be looked up by multiple keys at once. For example, Answers will typically be looked up by user_id and question_id.
3. Love the git practice of non-stop PRs, all day erryday. :+1: 
4. SHOULDA MATCHERS. :tada: 
5. More ActiveRecord validations please.

I am pretty happy with this.

File: Gemfile (hunk 2 above)
Comment (mikelikesbikes) [inline Gemfile]: or specify `group: :production`

File: app/assets/javascripts/admin/questions.js [hunk 20]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
+// All this logic will automatically be available in application.js.
```
Comment (mikelikesbikes) [inline app/assets/javascripts/admin/questions.js]: remove these.

File: app/assets/stylesheets/admin/questions.css [hunk 21]
```diff
@@ -0,0 +1,4 @@
+/*
+  Place all the styles related to the matching controller here.
+  They will automatically be included in application.css.
+*/
```
Comment (mikelikesbikes) [inline app/assets/stylesheets/admin/questions.css]: remove these.

File: app/controllers/answers_controller.rb [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+
+  def new
+    @answer = Answer.new
+  end
+
+  def create
+    @answer = Answer.new(params_answer)
+    if @answer.save
+      @answer.set_posted_at
+      redirect_to action: questions_path # Not sure if correct path?
```
Comment (mikelikesbikes) [inline app/controllers/answers_controller.rb]: action not needed... `redirect_to questions_path`

File: config/routes.rb [hunk 23]
```diff
@@ -1,4 +1,27 @@
...
+
+
+  resources :questions do
+    resources :responses
+    resources :answers do
+      resources :responses
+    end
+  end
+
+  resources :users, only: [:index, :show, :new]
```
Comment (mikelikesbikes) [inline config/routes.rb]: missing `:create`

File: app/models/answer.rb [hunk 24]
```diff
@@ -0,0 +1,8 @@
+class Answer < ActiveRecord::Base
+  include PostedAt
+
+  belongs_to :user
+  belongs_to :question
```
Comment (mikelikesbikes) [inline app/models/answer.rb]: missing validations for `user` and `question`

File: app/models/question.rb [hunk 25]
```diff
@@ -0,0 +1,15 @@
+class Question < ActiveRecord::Base
+  include PostedAt
+
+  belongs_to :user
```
Comment (mikelikesbikes) [inline app/models/question.rb]: validation.

File: app/models/question_tag.rb [hunk 26]
```diff
@@ -0,0 +1,4 @@
+class QuestionTag < ActiveRecord::Base
+  belongs_to :tag
+  belongs_to :question
```
Comment (mikelikesbikes) [inline app/models/question_tag.rb]: validations.

File: app/models/response.rb [hunk 27]
```diff
@@ -0,0 +1,7 @@
+class Response < ActiveRecord::Base
+  include PostedAt
+
+  belongs_to :respondable, polymorphic: true
```
Comment (mikelikesbikes) [inline app/models/response.rb]: validation

File: app/models/vote.rb [hunk 28]
```diff
@@ -0,0 +1,4 @@
+class Vote < ActiveRecord::Base
+  belongs_to :votable, polymorphic: true
+  belongs_to :user
```
Comment (mikelikesbikes) [inline app/models/vote.rb]: validations.

File: app/views/admin/users/_form.html.erb [hunk 29]
```diff
@@ -0,0 +1,35 @@
+<%= form_for @user, :url => edit_admin_user_path do |f| %>
```
Comment (mikelikesbikes) [inline app/views/admin/users/_form.html.erb]: rather than `@user` use `[:admin, @user]`

File: app/views/admin/users/show.html.erb [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+  <%= @user.bio %>
+</div>
+
+<div class = "user-questions">
+  <%= @user.questions.count%>
+  <ul>
+  <% @user.questions.each do |question|%>
+    <li><%= question.title %>
+    <%= question.votes.count %></li>
+  <% end %>
```
Comment (mikelikesbikes) [inline app/views/admin/users/show.html.erb]: <%= render @user.questions %>

File: app/views/answers/new.html.erb [hunk 31]
```diff
@@ -0,0 +1,3 @@
+<%= content_for :title, "New Answer" %>
```
Comment (mikelikesbikes) [inline app/views/answers/new.html.erb]: NICE!

File: db/migrate/20150114194348_create_questions.rb (hunk 11 above)
Comment (mikelikesbikes) [inline db/migrate/20150114194348_create_questions.rb]: when you create a migration use this:
```
t.references :user, index: true
```
if you've already distributed the migration use this:
```
add_index :questions, :user_id
```

File: db/migrate/20150114194931_create_answers.rb [hunk 32]
```diff
@@ -0,0 +1,13 @@
...
+      t.belongs_to :user
```
Comment (mikelikesbikes) [inline db/migrate/20150114194931_create_answers.rb]: when there's 2 foreign_keys use this:

```
add_index :answers, [:question_id, :user_id]
add_index :answers, :user_id
```

File: spec/models/answer_spec.rb [hunk 33]
```diff
@@ -0,0 +1,8 @@
+require 'rails_helper'
+
+RSpec.describe Answer, :type => :model do
+  it { should belong_to(:user)}
```
Comment (mikelikesbikes) [inline spec/models/answer_spec.rb]:
```ruby
let(:answer) { ... }
it { expect(answer).to belong_to(:user) }
```

File: db/migrate/20150114194931_create_answers.rb (hunk 32 above)
Comment (Nase00) [inline db/migrate/20150114194931_create_answers.rb]: this gave us "Index name 'index_answers_on_user_id' on table 'answers' already exists" and DB rollback, until we got rid of the line "add_index :answers, :user_id"

File: .travis.yml [hunk 34]
```diff
@@ -0,0 +1,8 @@
+language: ruby
+rvm:
+  - 2.0.0
+addons:
+  postgresql: "9.3"
+
+before_script:
+  - psql -c 'create database metaoverflow_test;' -U postgres
```
Comment (mikelikesbikes) [inline .travis.yml:8]: can you run "bundle exec rake db:test:prepare" instead, I dunno?

File: Gemfile [hunk 35]
```diff
@@ -27,14 +28,28 @@ gem 'sdoc', '~> 0.4.0',          group: :doc
...
+# gem 'factory_girl_rails'
+
+#Use rails_12factor
+gem 'rails_12factor', group: :production
+
+
+gem 'shoulda-matchers' , group: [:development, :test], require: false
+gem 'rspec-rails', group: [:development, :test]
+gem 'simplecov',group: [:development, :test], :require  => false
+gem 'ffaker', group: [:development, :test]
```
Comment (mikelikesbikes) [inline Gemfile:55]: nice touch.

File: app/assets/javascripts/voting.js [hunk 36]
```diff
@@ -0,0 +1,20 @@
+$('document').ready(function() {
+  function voteResponseHandler(response) {
```
Comment (mikelikesbikes) [inline app/assets/javascripts/voting.js:2]: take this and move it above the ready handler.

File: app/assets/stylesheets/admin/questions.css (hunk 21 above)
Comment (mikelikesbikes) [inline app/assets/stylesheets/admin/questions.css:4]: seriously.

File: app/controllers/admin/questions_controller.rb [hunk 37]
```diff
@@ -0,0 +1,2 @@
+class Admin::QuestionsController < ApplicationController
```
Comment (mikelikesbikes) [inline app/controllers/admin/questions_controller.rb:1]: shouldn't this extend from `Admin::ApplicationController`.

File: app/controllers/admin/users_controller.rb [hunk 38]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def show
+  end
+
+  def update
+    if @user.update(user_params)
+      redirect_to @user
+    else
+      render edit
```
Comment (mikelikesbikes) [inline app/controllers/admin/users_controller.rb:14]: this should be `:edit` not `edit`

File: app/controllers/answers_controller.rb [hunk 39]
```diff
@@ -0,0 +1,36 @@
...
+
+  def index
+  end
+
+  def new
+    @answer = Answer.new
+  end
+
+  def create
+    @answer = Answer.new(params_answer)
```
Comment (mikelikesbikes) [inline app/controllers/answers_controller.rb:14]:
``` ruby
@answer = current_user.answers.create(params_answer)
```

File: app/controllers/answers_controller.rb [hunk 40]
```diff
@@ -0,0 +1,36 @@
...
+  def set_answer
+    @answer = Answer.find(params[:id])
+  end
+
+  def set_question
+    @question = Question.find(params[:question_id])
+  end
+
+  def params_answer
+    params.require(:answer).permit(:content).merge(question_id: @question.id, user_id: current_user.id)
```
Comment (mikelikesbikes) [inline app/controllers/answers_controller.rb:34]: this is smelly. I'm not sure if it's bad. ... :smoking::fish:

File: app/controllers/responses_controller.rb [hunk 41]
```diff
@@ -0,0 +1,46 @@
...
+    @answer = Answer.find(params[:answer_id]) if params[:answer_id]
+  end
+
+  def set_question
+    @question = Question.find(params[:question_id])
+  end
+
+  def params_answer
+    if @answer
+      params.require(:response).permit(:content).merge(respondable_id: @answer.id, respondable_type: 'Answer', user_id: current_user.id)
```
Comment (mikelikesbikes) [inline app/controllers/responses_controller.rb:41]: shouldn't ever be setting the respondable_\* fields manually.

``` ruby
respondable = @answer ? @answer : @question

...merge(respondable: respondable, user: current_user)
```

File: app/helpers/votes_helper.rb [hunk 42]
```diff
@@ -0,0 +1,2 @@
+module VotesHelper
+end
```
Comment (mikelikesbikes) [inline app/helpers/votes_helper.rb:2]: remove.

File: app/models/question_tag.rb [hunk 43]
```diff
@@ -0,0 +1,6 @@
...
+
+  validates :tag_id, :question_id, presence: true
```
Comment (mikelikesbikes) [inline app/models/question_tag.rb:5]: separate validations to 1 per thing.

``` ruby
validates :tag, presence: true
validates :question, presence: true
```

File: app/models/user.rb [hunk 44]
```diff
@@ -0,0 +1,26 @@
...
+  validates :email, presence: true
+  validates_format_of :email,:with => /\A[^@\s]+@([^@\s]+\.)+[^@\s]+\z/
+  # validates_format_of :username, :without => /[^!@#$%^&*()]/
+  validates_format_of :username, :without => /\s/
+  has_secure_password
+
+  before_create :create_slug
+
+  def create_slug
+    self.slug = username.gsub(/\W/,"_")
```
Comment (mikelikesbikes) [inline app/models/user.rb:20]:
``` ruby
self.slug ||= username.gsub(/\W/,"_")
```

**Pull request closed without merging**